File: sdk/go/internal/signing/keystore.go
package signing

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"

	"golang.org/x/crypto/scrypt"

	"github.com/seocheon/sdk-go/internal/crypto"
)

// keystoreFile represents the structure of an encrypted keystore file.
// Compatible with the Web3 Secret Storage format (Ethereum-derived).
type keystoreFile struct {
	Crypto keystoreCrypto `json:"crypto"`
}

type keystoreCrypto struct {
	Cipher       string               `json:"cipher"`
	CipherText   string               `json:"ciphertext"`
	CipherParams keystoreCipherParams `json:"cipherparams"`
	KDF          string               `json:"kdf"`
	KDFParams    keystoreKDFParams    `json:"kdfparams"`
	MAC          string               `json:"mac"`
}

type keystoreCipherParams struct {
	IV string `json:"iv"`
}

type keystoreKDFParams struct {
	DKLen int    `json:"dklen"`
	N     int    `json:"n"`
	R     int    `json:"r"`
	P     int    `json:"p"`
	Salt  string `json:"salt"`
}

// loadAndDecryptKeystore reads a keystore file, derives the decryption key
// via scrypt, and decrypts the private key using AES-128-CTR.
func loadAndDecryptKeystore(path, passphrase string) (*crypto.PrivateKey, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading keystore file: %w", err)
	}

	var ks keystoreFile
	if err := json.Unmarshal(data, &ks); err != nil {
		return nil, fmt.Errorf("parsing keystore JSON: %w", err)
	}

	if ks.Crypto.KDF != "scrypt" {
		return nil, fmt.Errorf("unsupported KDF: %s (only scrypt is supported)", ks.Crypto.KDF)
	}
	if ks.Crypto.Cipher != "aes-128-ctr" {
		return nil, fmt.Errorf("unsupported cipher: %s (only aes-128-ctr is supported)", ks.Crypto.Cipher)
	}

	// Decode hex values
	salt, err := hex.DecodeString(ks.Crypto.KDFParams.Salt)
	if err != nil {
		return nil, fmt.Errorf("decoding salt: %w", err)
	}
	iv, err := hex.DecodeString(ks.Crypto.CipherParams.IV)
	if err != nil {
		return nil, fmt.Errorf("decoding IV: %w", err)
	}
	cipherText, err := hex.DecodeString(ks.Crypto.CipherText)
	if err != nil {
		return nil, fmt.Errorf("decoding ciphertext: %w", err)
	}
	mac, err := hex.DecodeString(ks.Crypto.MAC)
	if err != nil {
		return nil, fmt.Errorf("decoding MAC: %w", err)
	}

	// Derive key using scrypt
	dkLen := ks.Crypto.KDFParams.DKLen
	if dkLen == 0 {
		dkLen = 32
	}
	derivedKey, err := scrypt.Key([]byte(passphrase), salt, ks.Crypto.KDFParams.N, ks.Crypto.KDFParams.R, ks.Crypto.KDFParams.P, dkLen)
	if err != nil {
		return nil, fmt.Errorf("deriving key via scrypt: %w", err)
	}

	// Verify MAC: SHA256(derivedKey[16:32] + cipherText)
	macInput := append(derivedKey[16:32], cipherText...)
	calculatedMAC := sha256.Sum256(macInput)
	if !equalBytes(calculatedMAC[:], mac) {
		return nil, fmt.Errorf("MAC verification failed: incorrect passphrase or corrupted keystore")
	}

	// Decrypt using AES-128-CTR
	aesKey := derivedKey[:16]
	block, err := aes.NewCipher(aesKey)
	if err != nil {
		return nil, fmt.Errorf("creating AES cipher: %w", err)
	}

	stream := cipher.NewCTR(block, iv)
	privKeyBytes := make([]byte, len(cipherText))
	stream.XORKeyStream(privKeyBytes, cipherText)

	return crypto.NewPrivateKeyFromBytes(privKeyBytes)
}

// equalBytes compares two byte slices in constant-ish time.
func equalBytes(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	var diff byte
	for i := range a {
		diff |= a[i] ^ b[i]
	}
	return diff == 0
}

File: sdk/python/seocheon/infrastructure/signing/keystore.py
"""KeystoreSigningService: signs transactions using a local encrypted keystore file."""

from __future__ import annotations

import hashlib
import hmac
import json
import os

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

from seocheon.internal.crypto.address import address_from_pubkey
from seocheon.internal.crypto.keys import PrivateKey


class KeystoreSigningService:
    """Signs transactions using a local encrypted keystore file."""

    def __init__(self, keystore_path: str, passphrase_env: str) -> None:
        if not keystore_path or not passphrase_env:
            raise ValueError("keystore path and passphrase are required")
        self._priv_key = _load_and_decrypt_keystore(keystore_path, passphrase_env)
        self._pub_key = self._priv_key.pub_key()
        self._address = address_from_pubkey(self._pub_key)

    def sign(self, tx_bytes: bytes) -> bytes:
        return self._priv_key.sign(tx_bytes)

    def get_address(self) -> str:
        return self._address

    def get_pub_key(self) -> bytes:
        return self._pub_key


def _load_and_decrypt_keystore(path: str, passphrase: str) -> PrivateKey:
    """Read a keystore file, derive the decryption key via scrypt,
    and decrypt the private key using AES-128-CTR.
    """
    with open(path) as f:
        ks = json.load(f)

    crypto = ks["crypto"]
    if crypto["kdf"] != "scrypt":
        raise ValueError(f"unsupported KDF: {crypto['kdf']} (only scrypt is supported)")
    if crypto["cipher"] != "aes-128-ctr":
        raise ValueError(f"unsupported cipher: {crypto['cipher']} (only aes-128-ctr is supported)")

    kdf_params = crypto["kdfparams"]
    salt = bytes.fromhex(kdf_params["salt"])
    iv = bytes.fromhex(crypto["cipherparams"]["iv"])
    ciphertext = bytes.fromhex(crypto["ciphertext"])
    mac = bytes.fromhex(crypto["mac"])

    dk_len = kdf_params.get("dklen", 32)
    n = kdf_params["n"]
    r = kdf_params["r"]
    p = kdf_params["p"]

    # Derive key using scrypt
    derived_key = hashlib.scrypt(
        passphrase.encode("utf-8"),
        salt=salt,
        n=n,
        r=r,
        p=p,
        dklen=dk_len,
    )

    # Verify MAC: SHA256(derivedKey[16:32] + cipherText)
    mac_input = derived_key[16:32] + ciphertext
    calculated_mac = hashlib.sha256(mac_input).digest()
    if not hmac.compare_digest(calculated_mac, mac):
        raise ValueError("MAC verification failed: incorrect passphrase or corrupted keystore")

    # Decrypt using AES-128-CTR
    aes_key = derived_key[:16]
    cipher = Cipher(algorithms.AES(aes_key), modes.CTR(iv))
    decryptor = cipher.decryptor()
    priv_key_bytes = decryptor.update(ciphertext) + decryptor.finalize()

    return PrivateKey.from_bytes(priv_key_bytes)
